executor/brie.go

```diff
@@ -0,0 +1,464 @@
+		case ast.BRIEOptionSendCreds:
+			cfg.SendCreds = opt.UintValue != 0
+		}
+	}
+
+	if len(s.Tables) != 0 {
+		cfg.Filter.DoTables = make([]*filter.Table, 0, len(s.Tables))
+		for _, tbl := range s.Tables {
+			cfg.Filter.DoTables = append(cfg.Filter.DoTables, &filter.Table{
+				Name:   tbl.Name.O,
+				Schema: tbl.Schema.O,
+			})
+		}
+	}
+
+	switch s.Kind {
+	case ast.BRIEKindBackup:
+		e.backupCfg = &task.BackupConfig{Config: cfg}
+
+		for _, opt := range s.Options {
+			switch opt.Tp {
+			case ast.BRIEOptionLastBackupTS:
+				tso, err := b.parseTSString(opt.StrValue)
+				if err != nil {
+					b.err = err
+					return nil
+				}
+				e.backupCfg.LastBackupTS = tso
+			case ast.BRIEOptionLastBackupTSO:
+				e.backupCfg.LastBackupTS = opt.UintValue
+			case ast.BRIEOptionBackupTimeAgo:
+				e.backupCfg.TimeAgo = time.Duration(opt.UintValue)
+			case ast.BRIEOptionBackupTSO:
+				e.backupCfg.BackupTS = opt.UintValue
+			case ast.BRIEOptionBackupTS:
+				tso, err := b.parseTSString(opt.StrValue)
+				if err != nil {
+					b.err = err
+					return nil
+				}
+				e.backupCfg.BackupTS = tso
+			}
+		}
+
+	case ast.BRIEKindRestore:
+		e.restoreCfg = &task.RestoreConfig{Config: cfg}
+		for _, opt := range s.Options {
+			switch opt.Tp {
+			case ast.BRIEOptionOnline:
+				e.restoreCfg.Online = opt.UintValue != 0
+			}
+		}
+
+	default:
+		logutil.BgLogger().Fatal("unexpected BRIE statement kind", zap.Stringer("kind", s.Kind))
+		return nil
+	}
+
+	return e
+}
+
+// BRIEExec represents an executor for BRIE statements (BACKUP, RESTORE, etc)
+type BRIEExec struct {
+	baseExecutor
+
+	backupCfg  *task.BackupConfig
+	restoreCfg *task.RestoreConfig
+	info       *brieTaskInfo
+}
+
+// Open implements the Executor Open interface.
+func (e *BRIEExec) Open(ctx context.Context) error {
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return err
+	}
+
+	bq := globalBRIEQueue
+
+	e.info.connID = e.ctx.GetSessionVars().ConnectionID
+	e.info.queueTime = types.CurrentTime(mysql.TypeDatetime)
+	taskCtx, taskID := bq.registerTask(ctx, e.info)
+	defer bq.cancelTask(taskID)
+
+	progress, err := bq.acquireTask(taskCtx, taskID)
+	if err != nil {
+		return err
+	}
+	defer bq.releaseTask()
+
+	e.info.execTime = types.CurrentTime(mysql.TypeDatetime)
+	glue := &tidbGlueSession{se: e.ctx, progress: progress, info: e.info}
+
+	// manually monitor the Killed status...
+	go func() {
+		ticker := time.NewTicker(3 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				if atomic.LoadUint32(&e.ctx.GetSessionVars().Killed) == 1 {
+					bq.cancelTask(taskID)
+					return
+				}
+			case <-taskCtx.Done():
+				return
+			}
+		}
+	}()
+
+	switch e.info.kind {
+	case ast.BRIEKindBackup:
+		return handleBRIEError(task.RunBackup(taskCtx, glue, "Backup", e.backupCfg), ErrBRIEBackupFailed)
+	case ast.BRIEKindRestore:
+		return handleBRIEError(task.RunRestore(taskCtx, glue, "Restore", e.restoreCfg), ErrBRIERestoreFailed)
```

Comment:
I prefer to do this long-running process in the `Next` interface. The current execution framework puts query information into the `INFORMATION_SCHEMA.processlist` table (which is the underlying system table of `show processlist`) after the root executor `Open`s.

Reply:
For record, ChecksumTableExec also placed the long-running process in Open() rather than Next() (which I copied the design from). So we should review all long-running statements in the future.